File: LeetCode/Concurrency/1117. Building H2O.cpp
// 1117. Building H2O


// 1. Approach-01: Using Mutex, condition_variable, unique_lock
class H2O {
public:

    std::mutex mtx;
    std::condition_variable cv;
    int h;
    int o;

    H2O() {
        h = 0;
        o = 1;
    }

    void hydrogen(function<void()> releaseHydrogen) {
        
        std::unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, [this] { return h<2;});

        // releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen();

        h++;
        if(h==2) o = 0;
        cv.notify_all();
    }

    void oxygen(function<void()> releaseOxygen) {
        
        std::unique_lock<std::mutex> lck(mtx);
        cv.wait(lck, [this]{ return o==0;});

        // releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen();
        o = 1;
        h = 0;
        cv.notify_all();
    }
};

File: LeetCode/Concurrency/1117. Building H2O.java
// 1117. Building H2O
class H2O {

    //Approach-01: Using Semaphore
    Semaphore sm1, sm2;

    public H2O() {
        sm1 = new Semaphore(2, true);
        sm2 = new Semaphore(0, true);
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
		
        sm1.acquire();

        // releaseHydrogen.run() outputs "H". Do not change or remove this line.
        releaseHydrogen.run();
        
        sm2.release();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        
        sm2.acquire(2);

        // releaseOxygen.run() outputs "O". Do not change or remove this line.
		releaseOxygen.run();

        sm1.release(2);
    }
}

// 2. Using Synchronized

class H2O {

    //Approach-02: Using Synchronized
    private int h, o;

    public H2O() {
        h = 2;
        o = 0;
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
		
        synchronized(this){
            while(h<1) this.wait();

            // releaseHydrogen.run() outputs "H". Do not change or remove this line.
            releaseHydrogen.run();

            h--;
            if(h<1) o = 1;
            this.notifyAll();
        }
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        
        synchronized(this){
            while(o==0) this.wait();

             // releaseOxygen.run() outputs "O". Do not change or remove this line.
		    releaseOxygen.run();

            o--;
            h = 2;
            this.notifyAll();
        }
    }
}